02_knight_and_princess/matveev_alexey/task2.cpp

```diff
@@ -0,0 +1,99 @@
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+
+using VertexId = int;
+using EdgeId = int;
+
+struct Vertex {
+  const VertexId id;
+
+  explicit Vertex(const VertexId& _id) : id(_id) {}
+};
+
+struct Edge {
+  const EdgeId id;
+  const VertexId vertex_id1;
+  const VertexId vertex_id2;
+  Edge(const EdgeId& _id, VertexId _vertex_id1, VertexId _vertex_id2)
+      : id(_id), vertex_id1(_vertex_id1), vertex_id2(_vertex_id2) {}
+};
+
+class Graph {
+ public:
+  bool HasVertex(const VertexId& vertex_id) const {
+    for (const auto& vertex : vertexes_) {
+      if (vertex.id == vertex_id) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool AreConnected(const VertexId& _vertex_id1, const VertexId& _vertex_id2) {
+    assert(HasVertex(_vertex_id1) && "Vertex1 index is out of range");
+    assert(HasVertex(_vertex_id2) && "Vertex2 index is out of range");
+    if (_vertex_id1 == _vertex_id2) {
+      for (const auto& edge_id : connection_list_[_vertex_id1]) {
+        if (edges_[edge_id].vertex_id1 == _vertex_id1 &&
+            edges_[edge_id].vertex_id2 == _vertex_id1) {
+          return true;
+        }
+      }
+    } else {
+      for (const auto& edge_id : connection_list_[_vertex_id1]) {
+        if (edges_[edge_id].vertex_id1 == _vertex_id2 ||
+            edges_[edge_id].vertex_id2 == _vertex_id1) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  void AddVertex() {
+    connection_list_.insert({vertex_new_id_, std::vector<EdgeId>()});
+    vertexes_.emplace_back(GetNewVertexId());
```

Comment:
Плохая идея использовать сырое значение `id` до того момента, как ты вообще добавил объект. Нужно все делать последовательно. И вся логика по работе с `id` у тебя сосредоточена в функции `GetNewVertexId()`. Лучше будет так:
```suggestion
    const auto& new_vertex = vertexes_.emplace_back(GetNewVertexId());
    connection_list_.insert({new_vertex.id, std::vector<EdgeId>()});
```